core/src/main/scala/sttp/tapir/FieldName.scala

```diff
@@ -1,7 +1,27 @@
 package sttp.tapir
 
 case class FieldName(name: String, encodedName: String)
-
 object FieldName {
   def apply(name: String) = new FieldName(name, name)
 }
+
+case class FieldNames(fieldPath: List[FieldName]) {
+
+  def prependPath(f: FieldName): FieldNames =
+    copy(fieldPath = f :: fieldPath)
+
+  /** e.g. `user.address.street.name`.
+    */
+  val asPath: Option[FieldPath] =
+    fieldPath match {
+      case Nil => None
+      case l   => Some(FieldPath(l.map(_.encodedName).mkString(".")))
+    }
+}
+object FieldNames {
+  val empty: FieldNames = FieldNames(Nil)
+}
+
+case class FieldPath(value: String) extends AnyVal {
```

Comment:
I'm not sold on using a value class here. I like representing `FieldNames` as a value, though. Maybe it would be enough (and marginally less type safe) to simply have a `case class FieldPath(path: List[FieldName])`, with a `.toString` method rendering it in the dot-notation?

Reply:
Or rather `.toStringPath` as it needs to return an `Option[String]`

Reply:
I say yes and no and the same time :) 
I say yes because makes sense to have a `toStringPath` method and avoid the value class where reading the value.
But I say no(or at least I'd like to discuss it) because maybe it is less clear when you use it as parameter 

Here for example If we remove `FieldPath` we have an `Endo` of `String`, since this will be available to clients to build custom errors for custom validators maybe is better to keep the value class to make the signature more clear (?)
 
```scala
case class ValidationError(f: FieldPath => String, path: FieldNames = FieldNames.empty)
```

Reply:
Hm well I suppose this depends on how the other comments get resolved. Now that I understand more of your design, I agree that using a value class makes sense here, and `FieldPath => String` is definitely better than `String => String`.

But (there's always a but ;) ), as I mentioned in other comments, I would prefer to build the validation error messages externally - not as part of `ValidationResult/Error`s functionality. I think such a design makes more sense at it looses precision/information later, but I can also try to get convinced otherwise :)

Reply:
(and when we build error messages externally, I think there would be no `f` function which could cause confusion)